packages/pwa-kit-react-sdk/src/webpack/plugins.js

```diff
@@ -91,3 +95,104 @@ export class BuildMarkerPlugin {
+     * @param {Object[]} [options.customSchemas] - The object that include
+     * custom property schemas, you want to use this when you extend the schema.
+     * @param {String} [options.customSchemas.key] - The key of the property.
+     * @param {Boolean} [options.customSchemas.required] - Required.
+     * @param {Object} [options.customSchemas.schema] - The JSON schema for the
+     * custom property.
+     */
+    constructor(options = {}) {
+        this.path = options.path || './pwa-kit.config.json'
+        this.customSchemas = options.customSchemas || []
+    }
+
+    /**
+     * Called by webpack when this plugin is attached.
+     *
+     * @param compiler {Object} the webpack compiler
+     */
+    apply(compiler) {
+        compiler.hooks.compilation.tap('PwaKitConfigPlugin', (compilation) => {
+            try {
+                const config = this.getConfig(compiler)
+                this.validate(config)
+            } catch (e) {
+                compilation.errors.push(e)
+            }
+        })
+    }
+
+    /**
+     * Find the config file from project and return the value.
+     *
+     * @param compiler {Object} the webpack compiler
+     * @returns {object} - The configuration object
+     */
+    getConfig(compiler) {
+        const {fileSystem} = compiler.inputFileSystem
+        const absolutePath = path.resolve(this.path)
+        let configFile
+        try {
+            configFile = fileSystem.readFileSync(absolutePath)
+        } catch (e) {
+            throw new Error(`Missing PWA Kit config file, the file is required: ${absolutePath}`)
+        }
+        return JSON.parse(configFile)
+    }
+
+    /**
+     * Validate configurations based on pwa-kit-react-sdk schema
+     * and custom schema. Errors will be thrown when validation fails.
+     *
+     * @param config {Object} - the configuration object
+     */
+    validate(config) {
+        const ajv = new Ajv()
+        AjvMergePlugin(ajv)
+        ajv.addSchema(schema)
+        let valid
+
+        if (this.customSchemas.length) {
+            const required = [...schema.required]
+            const properties = this.customSchemas.reduce((prev, curr) => {
+                if (curr.required) {
+                    required.push(curr.key)
+                }
+                return {...prev, [curr['key']]: curr.schema}
+            }, {})
+            valid = ajv.validate(
+                {
+                    $merge: {
+                        source: {$ref: schema.$id},
+                        with: {
+                            required,
+                            properties
+                        }
+                    }
+                },
+                config
+            )
+        } else {
+            valid = ajv.validate(schema, config)
+        }
+
+        if (!valid) {
+            const message = ajv.errorsText(ajv.errors.filter((e) => e.schemaPath !== '#/$merge'), {
```

Comment:
What does this filter do?

Reply:
the `ajv-merge-patch` library always append a duplicated error in the error array, with a key `#/$merge`. I believe the purpose of that is to make user know that there are errors from the custom schema. That is not needed in our case, so i filtered out that one.